### PR TITLE
feat(nexus_deploy): Phase 1 Modul 1.2 — secret_sync.py

### PR DESCRIPTION
**Pull request by stefanko-ch**
## Summary

Phase 1 Modul 1.2 — migrate the two ~300-line bash heredoc blocks in `scripts/deploy.sh` that synced Infisical secrets into `stacks/{jupyter,marimo}/.infisical.env` to a single parametrised Python module.

Both blocks were functionally identical (only stack-name, paths, var-prefix differed). Collapsed into one `StackTarget` + one `render_remote_script` + one CLI invocation per stack.

**deploy.sh shrinks by 537 lines (5222 → 4685).** Cumulative Phase 1 delta: ~−883 lines (~16%).

## Behavioural parity

The Python rendering produces a bash script that mirrors deploy.sh's hardened logic byte-for-byte:
- POSIX shell-identifier key regex
- jq two-stage validation per folder (shape + TSV extraction)
- multi-line value guard with key-only warning (no secret-byte leakage)
- dotenv escape (`\\` → `\\\\`, `"` → `\"`)
- atomic same-directory `mktemp` + `mv`
- two outage gates (`succeeded==0`, `pushed==0`) leave existing file untouched
- legacy `.env` marker-block strip after successful write

Script is fed to `ssh nexus bash -s` via **stdin** (`_remote.ssh_run_script`), so the Infisical token never reaches argv / `ps` / CI logs.

## Eight rounds of hardening, eight named regression tests

One regression test per post-#495 hardening round in [`tests/unit/test_secret_sync.py`](tests/unit/test_secret_sync.py):

| # | Round | Test |
|---|---|---|
| R1 | `set -euo pipefail` first | `test_round_1_set_euo_pipefail_first_executable_line` |
| R2 | Credential transit (now stdin) | documented divergence in module docstring |
| R3 | `trap` cleans tmpfiles | `test_round_3_trap_cleans_all_tmpfiles` |
| R4 | jq two-stage validation | `test_round_4_two_stage_jq_validation` |
| R5 | Key regex `^[A-Za-z_][A-Za-z0-9_]*$` | `test_round_5_key_regex_inline` |
| R6 | Multi-line warning is key-only | `test_round_6_multiline_warning_does_not_emit_value` |
| R7 | Atomic write | `test_round_7_atomic_write_via_same_dir_mktemp` |
| R8 | Two outage gates | `test_round_8_two_outage_gates` |

Plus property-based dotenv-escape roundtrip via real `bash -c` eval, restart-on-change ordering, snapshot tests for the rendered jupyter + marimo scripts, and ten CLI integration tests covering the rc=0/1/2 dispatch contract.

## CLI contract

```
PROJECT_ID=… INFISICAL_TOKEN=… INFISICAL_ENV=dev GITEA_TOKEN=… \
    uv run python -m nexus_deploy secret-sync --stack <jupyter|marimo>
```

| rc | Meaning | deploy.sh handling |
|---|---|---|
| 0 | wrote new file (or no-touch outage gate fired cleanly) | continue |
| 1 | wrote, but ≥1 folder fetch failed | yellow warning, continue |
| 2 | transport / unexpected error | red, `exit 1` |

## Quality gates

- `ruff format --check`, `ruff check` — clean
- `mypy --strict` — clean
- `pytest` — 137/137 pass
- Coverage — **100%** on `src/nexus_deploy/secret_sync.py`, 100% project total
- `bash -n scripts/deploy.sh` — parses
- `shellcheck -S error scripts/deploy.sh` — clean

## Out of scope

- Restart-on-change runs in a separate `ssh_run` call after the sync (mirrors deploy.sh's two-step flow). Restart failures are logged but don't change the exit code.
- The `$` and backtick edge-case in dotenv values is documented as a deliberate parity choice — deploy.sh's sed escape doesn't neutralise them either, and they're vanishingly rare in real Infisical content. Tracked in the property-test docstring.
- Phase 3 (Modul 3.1) replaces the bash-rendering layer with paramiko + port-forwarded `requests`. Until then, `_remote.ssh_run_script` carries us.

## Test plan

- [ ] CI green on PR
- [ ] Manual spin-up on test VM with `enabled_services=jupyter,marimo`:
  - First run writes `.infisical.env` for both stacks; both containers come healthy
  - Second run produces byte-identical content → `docker compose up -d` is no-op (compose config-hash matched)
  - Outage simulation (revoke token mid-spin-up) → `wrote=0`, prior `.infisical.env` untouched

Closes part of #505 (Phase 1 sub-tasks 1.2.x).

### CODE REVIEW

File: src/nexus_deploy/secret_sync.py [hunk 1]
```diff
@@ -0,0 +1,461 @@
...
+        # Restart failure does NOT alter result.wrote; we just emit the
+        # error to stderr so the operator sees it.
+        restart_cmd = f"cd {shlex.quote(target.compose_dir)} && docker compose up -d {shlex.quote(target.name)}"
+        try:
+            run_cmd(restart_cmd)
+        except subprocess.CalledProcessError as exc:
+            # Class name only (no exc.cmd) — defence in depth: even
+            # though the restart command doesn't carry secrets, never
+            # let exception output reach the workflow log unfiltered.
+            print(f"  ⚠ docker compose up -d {target.name} failed ({type(exc).__name__})")
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py]: Fixed in 29d9d4d — restart failure now forwards the captured docker-compose stdout (indented, to stderr) so the operator can debug image-pull / compose-syntax / network errors. Mirrors deploy.sh:4909. exc.cmd is still NOT printed (defence in depth).

File: src/nexus_deploy/secret_sync.py [hunk 2]
```diff
@@ -0,0 +1,461 @@
...
+
+    # The bash below is deploy.sh:4616-4848 lifted near-verbatim.
+    # Differences from the heredoc form:
+    #   - Inputs come pre-decoded (shlex-quoted via stdin), no
+    #     base64 transit step (R2).
+    #   - Marker strings are interpolated as variables instead of
+    #     literal sed patterns, because sed-pattern-escaping the long
+    #     comment-style markers from Python is fragile. We use grep -F
+    #     + awk to find the marker line numbers and a one-pass cat
+    #     instead.
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py]: Fixed in 29d9d4d — comment now correctly describes the implementation: we use the same anchored sed range delete (`/^# === BEGIN nexus-secret-sync/,/^# === END nexus-secret-sync/d`) deploy.sh used. No grep -F / awk involved. The matched prefix has no regex specials, so no Python-side escaping is needed; only the variable parts (Jupyter / Marimo wording) sit AFTER the matched prefix and only matter for the printf that writes the new block.

File: src/nexus_deploy/__main__.py [hunk 3]
```diff
@@ -178,8 +180,146 @@ def _infisical_bootstrap(args: list[str]) -> int:
...
+    Exit codes (deploy.sh's case-block dispatches):
+    - 0: success, OR sync correctly chose not to write (one of the
+         two outage gates fired — operator sees a stderr warning,
+         existing file untouched, deploy.sh continues)
+    - 1: partial — file written but at least one folder fetch failed
+         (deploy.sh-side: warn-and-continue; the operator can fix the
+         offending folder via the Infisical UI without aborting)
+    - 2: hard failure — invalid `--stack`, missing required env,
+         transport (ssh) failure, no parseable RESULT line from the
+         remote script, unexpected exception. deploy.sh-side: abort.
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/__main__.py]: Fixed in 29d9d4d — docstring now matches the actual rc=0 (warn-and-continue) semantic for the no-RESULT case, mirroring deploy.sh's `[ -z "$JUP_PUSHED" ]` branch. The inner script's own stderr is already in the workflow log for diagnosis; aborting would be over-strict here.

File: src/nexus_deploy/__main__.py [hunk 4]
```diff
@@ -178,8 +180,146 @@ def _infisical_bootstrap(args: list[str]) -> int:
...
+        # which could embed secret-bearing values.
+        print(
+            f"secret-sync: unexpected error ({type(exc).__name__})",
+            file=sys.stderr,
+        )
+        return 2
+
+    # No parseable RESULT line — remote script broke in an unhandled
+    # way. Treat as hard failure (rc=2) so deploy.sh aborts; the script
+    # stdout/stderr is already in the workflow log for diagnosis.
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/__main__.py]: Fixed in 29d9d4d — inline comment now says rc=0 with the correct rationale (matches the all-zeros outcome from either no-RESULT or the legitimate jq-missing path; both demand warn-and-continue, not abort).

File: src/nexus_deploy/secret_sync.py [hunk 5]
```diff
@@ -0,0 +1,473 @@
...
+SECRETS_URL={secrets_url_q}
+
+CFG=$(mktemp)
+SEEN=$(mktemp)
+APPEND=$(mktemp)
+NEW_BLOCK=$(mktemp)
+TSV=$(mktemp)
+TMP_OUT=""
+chmod 600 "$CFG" "$APPEND" "$NEW_BLOCK" "$TSV"
+trap 'rm -f "$CFG" "$SEEN" "$APPEND" "$NEW_BLOCK" "$TSV" "$TMP_OUT"' EXIT
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py]: Fixed in fcb742a — trap now guards the optional tmpfiles ([-n "$TMP_OUT"] && rm). `rm -f ""` is silent on macOS/GNU coreutils with -f, but explicit > implicit. Also added a trailing `true` so the trap's own exit doesn't leak the [-n] check result.

File: src/nexus_deploy/secret_sync.py [hunk 6]
```diff
@@ -0,0 +1,473 @@
...
+cat "$NEW_BLOCK" >> "$TMP_OUT"
+mv "$TMP_OUT" "$ENV_FILE"
+chmod 600 "$ENV_FILE"
+WROTE=1
+
+if [ -f "$LEGACY_ENV" ]; then
+    LEGACY_TMP=$(mktemp -p "$(dirname "$LEGACY_ENV")" .env.XXXXXX)
+    chmod 600 "$LEGACY_TMP"
+    sed '/^# === BEGIN nexus-secret-sync/,/^# === END nexus-secret-sync/d' "$LEGACY_ENV" > "$LEGACY_TMP"
+    mv "$LEGACY_TMP" "$LEGACY_ENV"
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py:367]: Fixed in fcb742a — LEGACY_TMP is now declared empty at the top, included in the EXIT trap with a non-empty guard, and reset to empty after the successful mv (so a clean run doesn't try to rm a path consumed by the rename). Interrupt between mktemp and mv now triggers cleanup.

File: src/nexus_deploy/secret_sync.py [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+    #     delete as deploy.sh (`/^# === BEGIN nexus-secret-sync/,
+    #     /^# === END nexus-secret-sync/d`). The match is on the
+    #     fixed `# === BEGIN/END nexus-secret-sync` prefix only —
+    #     no per-stack interpolation — so no Python-side regex
+    #     escaping is needed. The variable parts of the marker text
+    #     (Jupyter / Marimo wording) sit AFTER the matched prefix
+    #     and only matter for the `printf` that writes the new block.
+    #   - The "BEGIN marker" comment text in the new block matches
+    #     deploy.sh's wording exactly (target.begin_marker).
+    return f"""set -euo pipefail
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py]: Fixed in 1bf4203 — replaced 'deploy.sh:4616-4848' with a conceptual reference to the legacy heredoc (history pre-#510). Same for the other ~10 line-number citations in the module + tests.

File: tests/unit/test_secret_sync.py [hunk 8]
```diff
@@ -0,0 +1,747 @@
...
+    target = StackTarget(name="marimo")
+    assert target.env_file == "/opt/docker-server/stacks/marimo/.infisical.env"
+    assert target.legacy_env_file == "/opt/docker-server/stacks/marimo/.env"
+    assert target.compose_dir == "/opt/docker-server/stacks/marimo"
+
+
+def test_stack_target_begin_marker_capitalised() -> None:
+    """Marker comment includes capitalised stack name (mirrors deploy.sh:4811/4925)."""
+    assert "Infisical → Jupyter env" in StackTarget(name="jupyter").begin_marker
+    assert "Infisical → Marimo env" in StackTarget(name="marimo").begin_marker
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply tests/unit/test_secret_sync.py:55]: Fixed in 1bf4203 — same theme: replaced all 'deploy.sh:4811/4925' / 'deploy.sh:4727' / 'deploy.sh:4761' / 'deploy.sh:4904' / 'deploy.sh:4909' citations with conceptual references that stay accurate as deploy.sh shrinks in later phases.

File: scripts/deploy.sh [hunk 9]
```diff
@@ -4552,544 +4552,48 @@ REMOTE_TRIG_EOF
...
-#     only" workflow), and saves a needless second recreate cycle
-#     here when the secret set hasn't changed.
-#
-# In notebooks: `import os; os.environ['R2_ACCESS_KEY']`. GITEA_TOKEN
-# included as a special case (same as Kestra) — auto-generated each
-# spin-up, not in Infisical.
-# --------------------------------------------------------------------------
+# Migrated from a 358-line bash heredoc to nexus_deploy.secret_sync (#505).
+# Exit-code contract enforced by the Python CLI:
+#   0 = wrote new .infisical.env (or no-touch outage gate fired cleanly)
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply scripts/deploy.sh]: Fixed in 1bf4203 — deploy.sh's inline exit-code summary now spells out the three rc=0 sub-cases (wrote / outage-gate / no-RESULT soft no-op), matching the docstring in nexus_deploy.__main__._secret_sync.

File: src/nexus_deploy/secret_sync.py [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+import subprocess
+import sys
+from collections.abc import Callable
+from dataclasses import dataclass
+
+from nexus_deploy import _remote
+
+# Marker block delimiters. Must match deploy.sh:4811-4813 (Jupyter) and
+# the equivalent Marimo lines BYTE-FOR-BYTE — operators expect the same
+# greppable marker, and the legacy-env strip step uses the same regex.
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py]: Fixed in 1bf4203 — marker-block comment now references the legacy deploy.sh secret-sync conceptually instead of citing line numbers that no longer exist post-#510.

File: src/nexus_deploy/secret_sync.py [hunk 11]
```diff
@@ -0,0 +1,484 @@
...
+        target=target,
+        project_id=project_id,
+        infisical_token=infisical_token,
+        infisical_env=infisical_env,
+        gitea_token=gitea_token,
+    )
+
+    completed = run_script(script)
+    result = parse_result(completed.stdout)
+    if result is None:
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/secret_sync.py:457]: Fixed in 3be26dc — `run_sync_for_stack` now forwards remote stdout (minus the RESULT line) to local sys.stderr, restoring the deploy.sh-heredoc UX where remote warnings stream directly to the workflow log. Used merge_stderr=True + line-strip rather than merge_stderr=False to avoid changing the _remote.ssh_run_script default contract; the wire-format RESULT line is recognisable enough by its prefix that splitting is safe. Regression test added: `test_run_sync_forwards_remote_warnings_to_local_stderr`.

File: src/nexus_deploy/__main__.py [hunk 12]
```diff
@@ -178,8 +181,149 @@ def _infisical_bootstrap(args: list[str]) -> int:
...
+            f"secret-sync: {stack} wrote {result.pushed} env-vars "
+            f"({result.collisions} cross-folder collision(s) — first-wins applied)",
+        )
+        return 0
+    # wrote=False with non-zero counters — one of the two outage gates
+    # fired (succeeded==0 or pushed==0). Existing file untouched.
+    # Operator already saw the cause from the inner script's stderr.
+    print(
+        f"secret-sync: {stack} skipped .infisical.env update (kept previous; see prior warning)",
+    )
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply src/nexus_deploy/__main__.py:321]: Fixed in 3be26dc — same root-cause as the secret_sync.py finding: remote diagnostics are now forwarded to local stderr in run_sync_for_stack. The 'see prior warnings' messages now point at actual visible warnings.

File: scripts/deploy.sh [hunk 13]
```diff
@@ -4552,544 +4552,50 @@ REMOTE_TRIG_EOF
...
-#
-# In notebooks: `import os; os.environ['R2_ACCESS_KEY']`. GITEA_TOKEN
-# included as a special case (same as Kestra) — auto-generated each
-# spin-up, not in Infisical.
-# --------------------------------------------------------------------------
+# Migrated from a 358-line bash heredoc to nexus_deploy.secret_sync (#505).
+# Exit-code contract enforced by the Python CLI:
+#   0 = wrote new .infisical.env, OR no-touch outage gate fired cleanly,
+#       OR remote script produced no parseable RESULT line (soft no-op,
+#       inner script's own stderr is already in the workflow log)
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply scripts/deploy.sh:4561]: Fixed in 3be26dc — Python implementation now forwards remote stderr/stdout to local stderr (legacy deploy.sh heredoc UX restored), so the deploy.sh comment's claim about 'inner script's own stderr is already in the workflow log' is now accurate.

File: tests/unit/test_secret_sync.py [hunk 14]
```diff
@@ -0,0 +1,780 @@
...
+def test_escape_dotenv_value_basic() -> None:
+    """Round 6 escape rules — mirror legacy deploy.sh secret-sync sed."""
+    assert escape_dotenv_value("plain") == "plain"
+    assert escape_dotenv_value('with"quote') == 'with\\"quote'
+    assert escape_dotenv_value("with\\backslash") == "with\\\\backslash"
+    # Order matters: backslash escape FIRST, then quote
+    assert escape_dotenv_value('mix"\\') == 'mix\\"\\\\'
+
+
+@given(st.text(alphabet=st.characters(blacklist_characters="\n\r\x00$`"), max_size=40))
```
[the remark this answers is not in the document]
Comment (stefanko-ch) [reply tests/unit/test_secret_sync.py]: Fixed in 683a81c — renamed to `exclude_characters` (the new name is already supported in hypothesis 6.152.4; this avoids the eventual DeprecationWarning that our filterwarnings=["error"] config would turn into a hard test failure).
